src/tensor/safetensors.rs

```diff
@@ -0,0 +1,143 @@
+    data: Vec<u8>,
+}
+
+#[derive(Default)]
+pub struct Writer {
+    tensors: BTreeMap<String, TensorData>,
+}
+
+impl Writer {
+    pub fn new() -> Self {
+        let tensors = BTreeMap::new();
+        Self { tensors }
+    }
+
+    pub fn add<S, T, E: Dtype + SafeDtype, D: DeviceStorage + CopySlice<E>>(
+        mut self,
+        key: String,
+        tensor: Tensor<S, E, D, T>,
+    ) -> Self
+    where
+        S: Shape,
+        <D as DeviceStorage>::Storage<S, E>: HasUnitType<Unit = E> + AsVec,
+    {
+        let dtype = E::safe_dtype();
+        let shape = tensor.shape().concrete().into();
+        let data = tensor.as_vec();
+        let data: Vec<u8> = data.iter().flat_map(|f| f.to_le_bytes()).collect();
+        let tdata = TensorData { dtype, shape, data };
+        self.tensors.insert(key, tdata);
+        self
+    }
+
+    pub fn save(&self, path: &Path) -> Result<(), SafeTensorError> {
+        let views: BTreeMap<String, TensorView> = self
+            .tensors
+            .iter()
+            .map(|(k, tensor)| {
+                (
+                    k.clone(),
+                    TensorView::new(tensor.dtype, tensor.shape.clone(), &tensor.data),
+                )
+            })
+            .collect();
+        serialize_to_file(&views, &None, path)
+    }
+}
+
+pub trait SafeDtype: Sized {
+    type Array: IntoIterator<Item = u8>;
+
+    fn from_le_bytes(bytes: &[u8], index: usize) -> Self;
+    fn to_le_bytes(self) -> Self::Array;
+    fn safe_dtype() -> SDtype;
+}
+
+impl SafeDtype for f32 {
+    type Array = [u8; 4];
+    fn from_le_bytes(bytes: &[u8], index: usize) -> Self {
+        Self::from_le_bytes([
+            bytes[index],
+            bytes[index + 1],
+            bytes[index + 2],
+            bytes[index + 3],
+        ])
+    }
+
+    fn to_le_bytes(self) -> Self::Array {
+        self.to_le_bytes()
+    }
+
+    fn safe_dtype() -> SDtype {
+        SDtype::F32
+    }
+}
+
+impl SafeDtype for f64 {
+    type Array = [u8; 8];
+    fn from_le_bytes(bytes: &[u8], index: usize) -> Self {
+        Self::from_le_bytes([
+            bytes[index],
+            bytes[index + 1],
+            bytes[index + 2],
+            bytes[index + 3],
+            bytes[index + 4],
+            bytes[index + 5],
+            bytes[index + 6],
+            bytes[index + 7],
+        ])
+    }
+
+    fn safe_dtype() -> SDtype {
+        SDtype::F64
+    }
+
+    fn to_le_bytes(self) -> Self::Array {
+        self.to_le_bytes()
+    }
+}
+
+impl<S: Shape, E: Dtype + SafeDtype, D: DeviceStorage + CopySlice<E>, T> Tensor<S, E, D, T> {
+    pub fn safetensors_writer() -> Writer {
+        Writer::new()
+    }
+
+    pub fn load<'a>(
+        &mut self,
+        tensors: &SafeTensors<'a>,
+        key: &str,
+    ) -> Result<(), SafeTensorError> {
+        let tensor = tensors.tensor(key)?;
+        let v = tensor.data();
+        let num_bytes = std::mem::size_of::<E>();
+        if (v.as_ptr() as usize) % num_bytes == 0 {
```

Comment:
How sure are you that this is correct? Is this expected use for safetensors? I see std::ptr has is_aligned_to but its on nightly unfortunately.

Reply:
Gotcha, makes sense, we can keep for now and then move to using is_aligned_to whenever it becomes stable